docs/05-engineering/git-gitlab-github-bitbucket.md

```diff
@@ -1,29 +1,144 @@
-# Git, GitLab, GitHub, and Bitbucket
+# Git
 
 ## Git Overview
 
-As you probably already know, Git is a source code versioning system that lets you locally track changes and push or pull changes from remote resources. Many projects use Git. Here are some references to review if you are unfamiliar with it.
+As you probably already know, Git is a source code versioning system that lets you locally track changes and push or pull changes from remote resources. All projects use Git. Here are some references to review if you are unfamiliar with it.
 
 - <https://git-scm.com/>
 - <https://help.github.com/articles/git-and-github-learning-resources/>
 - Getting Git Right <https://www.atlassian.com/git/>
 - Version Control with Git book <http://www.amazon.com/Version-Control-Git-collaborative-development/dp/1449316387>
 - Learn "how git works" at a lower level, and some useful stuff like rebasing (workflow and interactive), reflog, etc. <https://www.youtube.com/watch?v=MYP56QJpDr4>
-- if you want a non-video companion to the first half, <http://git-scm.com/book/en/v2/Git-Internals-Git-Objects> is a great place to start.
+- If you want a non-video companion to the first half, <http://git-scm.com/book/en/v2/Git-Internals-Git-Objects> is a great place to start.
 
-## What's the difference between GitLab, GitHub, and Bitbucket?
+## Git best practices
+
+_Note: your project may include additional or alternate code review practices - please review your project developer/contributor documentation!_
+
+- Atomic commits:
+  - Generally commits should be as small as possible.
+  - The code should be in a functional state after each commit (even if incomplete) - avoid committing code which will won't run at all.
+- Write good commit messages. A good commit message should:
+  - Reference the associated ticket ID using a prefix, e.g. `ABC-123:`.
+  - Have a one line (max 80 char) title/summary of the change, followed by a blank line and then further detail which should:
+    - Explain the reason(s) for the change.
+    - Detail what changed and any consequences of this (on code, data, users, security etc) - include any key terms that would help a future person searching commits.
+    - Often this will include the same detail that you might include in code comments.
+- Feature branches, pull requests (PR) & merge requests (MR):
+  - These should contain a closely related series of commits that solves a single problem - avoid including changes that are unrelated.
+  - Most often a single (well scoped) ticket will be resolved in a single PR/MR.
+    - However it is good to break larger changes over multiple PRs/MRs that each addresses a logical piece of the whole - this can make the code easier to review.
+  - When merging PRs and MRs we prefer a fast-forward/rebase (rather than a merge commit or squash) since this retains a clean and linear commit history.
```

Comment:
I am a big fan of squash merges. If you are being good about having one idea/feature/etc per PR, it makes sense for you to have a single commit with a good message in the main branch once it is squashed merged.

Reply:
Yeah, I think for small PRs squashing is certainly a reasonable option - I'll add a note to that effect. I think for bigger ones keeping the more granular commit history available is still a helpful default though, as it often contains details on intention around smaller changes that would be hard to summarize in a squash commit without the commit context.